src/types/streaming_server/request.rs

```diff
@@ -0,0 +1,323 @@
+pub struct CreateTorrentBlobBody {
+    pub blob: String,
+}
+
+/// # Examples
+///
+/// Example which creates a request url with body for the server:
+/// `http://127.0.0.1:11470/df389295484b3059a4726dc6d8a57f71bb5f4c81/1?tr=https%3A%2F%2Fexample.com%2Fmy-awesome-video.mp4`
+///
+/// ```
+/// use stremio_core::types::streaming_server::{CreateTorrentRequest, CreatedTorrent};
+///
+/// let request: http::Request<CreatedTorrent> = CreateTorrentRequest {
+///     server_url: "http://127.0.0.1:11470/".parse().unwrap(),
+///     sources: vec!["https://example.com/my-awesome-video.mp4".into()],
+///     info_hash: "df389295484b3059a4726dc6d8a57f71bb5f4c81".parse().unwrap(),
+///     file_idx: 1,
+/// }.into();
+///
+/// assert_eq!("http://127.0.0.1:11470/df389295484b3059a4726dc6d8a57f71bb5f4c81/1?tr=https%3A%2F%2Fexample.com%2Fmy-awesome-video.mp4", request.uri().to_string());
+///  // TODO: assert_eq!(&(), request.body());
+/// ```
+pub struct CreateTorrentRequest {
+    pub server_url: Url,
+    pub sources: Vec<String>,
+    pub info_hash: InfoHash,
+    pub file_idx: u64,
+}
+
+impl From<CreateTorrentRequest> for Request<CreatedTorrent> {
+    fn from(val: CreateTorrentRequest) -> Self {
+        let url = {
+            let mut uri = val
+                .server_url
+                .join(&format!("{}/{}", val.info_hash, val.file_idx))
+                .expect("Should always be valid Url");
+
+            {
+                let mut x = uri.query_pairs_mut();
+                for source in &val.sources {
+                    x.append_pair("tr", source);
+                }
+            }
+
+            uri
+        };
+
+        let create_torrent = CreatedTorrent {
+            torrent: super::Torrent {
+                info_hash: val.info_hash,
+            },
+            peer_search: PeerSearch::new(40, 200, val.info_hash, val.sources),
+            guess_file_idx: None,
+        };
+
+        Request::builder()
+            .uri(url.as_str())
+            .method("POST")
+            .header(CONTENT_TYPE, "application/json")
+            .body(create_torrent)
+            .expect("Should always be valid Request!")
+    }
+}
+
+pub struct CreateMagnetRequest {
+    pub server_url: Url,
+    pub info_hash: InfoHash,
+    pub announce: Vec<String>,
+}
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CreateMagnetBody {
+    pub torrent: CreateMagnetTorrent,
+    pub peer_search: Option<PeerSearch>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CreateMagnetTorrent {
+    pub info_hash: InfoHash,
+}
+
+impl From<CreateMagnetRequest> for Request<CreateMagnetBody> {
+    fn from(val: CreateMagnetRequest) -> Self {
+        let info_hash = val.info_hash;
+
+        let body = CreateMagnetBody {
+            torrent: CreateMagnetTorrent {
+                info_hash: val.info_hash.to_owned(),
+            },
+            peer_search: if !val.announce.is_empty() {
+                Some(PeerSearch {
```

Comment:
Why not use `PeerSearch::new` here?

Reply:
the `new` function was created after the fact.
Updated!